File: readwrite.go
package main

import (
"fmt"
"math/rand"
"time"
)

var master = []int{0,1}
var c = make(chan int,1)
var count = 0
func writer() {
	<-c
	fmt.Printf("The first value of the array before writing is %d\n",master[0])
	master[0] += 1
	fmt.Printf("The second value of the array before writing is %d\n",master[1])
	master[1] += 1
	c<-1
		time.Sleep(time.Duration(rand.Intn(300)) * time.Millisecond)

	
}

func reader(id string) {
	if(count < 1){<-c}
	count += 1
	fmt.Printf("reader %s sees %d  as the value for element 1 \n",id,master[0])
	fmt.Printf("reader %s sees %d  as the value for element 2 \n",id,master[1])
	count -= 1 
	if(count < 1){c<-1}
	time.Sleep(time.Duration(300000) * time.Millisecond)

}
func main() {
rand.Seed( time.Now().UTC().UnixNano())
c <- 1
for i := 0; i < 10; i++ {
	go reader("1")
	go writer()
	go reader("2")
	go reader("3")
	go reader("4")
	}
time.Sleep(time.Duration(1000) * time.Millisecond)
}

File: readwrite.py
#readwrite.py

from random import randint
import time
import threading


master = [0,1]
count = 0
def write():
	for x in xrange(1,10):
		sema.acquire()
		print "The first value of the array before writing is " + str(master[0])
		master[0] += 1
		print "The second value of the array before writing is " + str(master[1])
		master[1] +=1
		x+=1
		sema.release()
		time.sleep(1)


def reader(id):
	global count
	for x in xrange(1,10):
		if( count<1):
			sema.acquire()
		count +=1
		print "reader " + str(id) + " sees " + str(master[0]) + " as the first value of the array\n"
		print "reader " + str(id) + " sees " + str(master[1]) + " as the second value of the array\n"
		
		x+=1
		count -=1
		if(count<1):
			sema.release()
		time.sleep(2)


sema = threading.Semaphore()
writer = threading.Thread(target=write)
reader1 = threading.Thread(target=reader, args=(1,));
reader2 = threading.Thread(target=reader, args=(2,));
reader3 = threading.Thread(target=reader, args=(3,));
reader4 = threading.Thread(target=reader, args=(4,));

writer.start()
reader1.start()
reader2.start()
reader3.start()
reader4.start()
writer.join()
reader1.join()
reader2.join()
reader3.join()
reader4.join()



